content/2022/simulacija-muve.md

```diff
@@ -2,3 +2,147 @@
+
+Fitnes funkcija je funkcija koju implementira korisnik koja treba da odredi koliko je dobar svaki gen tj. mreža za određeni zadatak. Ona će biti pozvana jednom za svaku generaciju. biće joj prosleđena lista gena. Korisnik pomoću svakog gena može da generiše neuralnu mrežu koristeći funkciju koju smo implementirali. Fitnes funkcija treba da promeni fitness član genoma u listi za svaki genom prema želji korisnika. Korisnik može da odabere na koji način će oceniti svaku mrežu. Kako bi koristio mreže korisnik može na objektu mreže koji je inicijalizovao pomoću gena da pozove funkciju Activate i da joj prosledi ulazne podatke za mrežu formata NumPy matrice. U svakoj koloni matrice je zaseban set ulaznih podataka mreže (input features). Imamo više kolona u slučaju da kroz istu mrežu želimo odjednom da prosledimo više setova aktivacija. U našem slučaju imaćemo samo jednu kolonu jer obrađujemo samo ulazne podatke od jedne bube za svaku mrežu. Funkcija Activate nam vraća vektor sličnog formata kao na ulazu sa izlaznim vrednostima mreže za svaki set.
+
+Označimo sa $X_{j}^{i}$ j-ti ulazni podatak (input feature) od i-tog seta ulaznih podataka. Onda format matrice ulaznih podataka za funkciju Activate izgleda ovako:
+
+| Set 1 | Set 2 | ... | Set m    |
+| ----- | ----- | ----- | ----- |
+| $X_{1}^{1}$ | $X_{1}^{2}$ | ... | $X_{1}^{m}$ |
+| $X_{2}^{0}$ | $X_{2}^{2}$ | ... | $X_{2}^{m}$ |
+| ... | ... | ... | ... |
+| $X_{n}^{1}$ | $X_{n}^{2}$ | ... | $X_{n}^{m}$ |
+
+Ovde imamo n ulaznih podataka za svaki set. Imamo m setova. Ovo znači da naša mreža u prvom sloju ima n neurona i da će u ovom slučaju obrađivati m setova ulaznih podataka paralelno. Vrednost m ne mora da bude ista za svaki poziv Activate funkcije dok n mora jer je već definisana arhitekturom mreže.
+
+**bitno je voditi računa o orijentacijama vektora u implementaciji!**
+
+#### Funkcionisanje genetskog algoritma
+Sve informacije o nekoj mreži nam se nalaze u njenom genu. Objekat gena sadrži razne podatke o mreži u sebi, kao i sve parametre mreže. Pri pravljenju populacije svi parametri dobijaju nasumične vrednosti po gausovoj (normalnoj) raspodeli sa standardnom devijacijom koja se prosleđuje konstruktoru.
+
+Algoritam se sastoji iz 3 faze
+ - selection
+ - crossover
+ - mutation
+
+##### Selection faza:
+Nakon što se završi fitnes funkcija program bira najboljih 10% gena. Od njih pravi 'mating pool'.
+
+##### Crossover faza:
+Novi geni se generišu tako što se biraju nasumično dva 'rotitelja' iz 'mating poola'. Parametri nove mreže mogu da se dobiju na razne načina. Konkretno u našem slučaju za svaki parametar mreže ima $x$ šanse da bude iz prvog roditelja i $1-x$ šanse da bude iz drugog ($x$ je jedan od parametara koji prosleđujemo run funkciji).
+
+##### Mutation faza:
+Nakon ovog dela vrše se nasumične mutacije na svakom od novih gena. Učestalost ovih mutacija takođe može da se prosledi u run funkciji.
```

Comment:
- Šta znači "ovog dela", jel to mutaciona faza ili ukrštanje?
- Objasniti šta znači mutacija u ovom konkretnom slučaju
- Objasniti koji problem rešava mutacija